src/threading/cthread.cpp

```diff
@@ -0,0 +1,191 @@
+#include "cthread.hpp"
+#include <cassert>
+
+using namespace nest::mc::threading::impl;
+
+// RAII owner for a task in flight
+struct task_pool::run_task {
+    task_pool& pool;
+    lock& lck;
+    task tsk;
+
+    run_task(task_pool&, lock&);
+    ~run_task();
+};
+
+// Own a task in flight
+// lock should be passed locked,
+// and will be unlocked after call
+task_pool::run_task::run_task(task_pool& pool, lock& lck):
+    pool{pool},
+    lck{lck},
+    tsk{}
+{
+    std::swap(tsk, pool._tasks.front());
+    pool._tasks.pop_front();
+    
+    lck.unlock();
+    pool._tasks_available.notify_all();
+}
+
+// Release task
+// Call unlocked, returns unlocked
+task_pool::run_task::~run_task() {
+    lck.lock();
+    tsk.second->in_flight--;
+    
+    lck.unlock();
+    pool._tasks_available.notify_all();
+}
+
+// runs forever until quit is true
+void task_pool::run_tasks_loop() {
+    lock lck = {_tasks_mutex, std::defer_lock};
+    while (true) {
+        lck.lock();
+
+        while (! _quit && _tasks.empty()) {
+            _tasks_available.wait(lck);
+        }
+        if (_quit) {
+            return;
+        }
+    
+        run_task run{*this, lck};
+        run.tsk.first();
+    }
+}
+
+// run until out of tasks for a group
+void task_pool::run_tasks_while(task_group* g) {
+    lock lck{_tasks_mutex, std::defer_lock};
+    while (true) {  
+        lck.lock();
+
+        while (! _quit && g->in_flight && _tasks.empty()) {
+            _tasks_available.wait(lck);
+        }
+        if (_quit || ! g->in_flight) {
+            return;
+        }
+
+        run_task run{*this, lck};
+        run.tsk.first();
+    }
+}
+
```

Comment:
It would make (if I understand the code correctly) `run_task_loops()` equivalent to `run_tasks_while(nullptr)`; they are both doing the same thing modulo a different exit condition, which can be wrapped up in a method like `finished()`.